### ISSUE TITLE
調査: non-playing競りAIのスート選択が♠/♥へ極端に偏る原因を特定する

### ISSUE THREAD

**Issue by eletim** [eletim/napoleon-web#246]
Related: #202, #244
Base branch: `feature/issue-202-non-playing-ai`

## 背景

RuleBased / Conservative 50:50 のFrozen競り相手、Reward v3、相対reward、all-pass無効試合（全員0）の条件で non-playing PPO を 100 iterations × 500 logical games/iteration 学習した。

最終 iteration の candidate bidding diagnostics は以下だった。

- passRate: 0.2492
- candidateNapoleonFormationRate: 0.5932
- declarationSuccessRate: 0.3533
- bid target distribution:
  - 13: 1776
  - 14: 1204
  - 15: 221
  - 16: 34
  - 17: 8
  - 18: 25
  - 19: 13
- suit distribution:
  - spades: 1834
  - hearts: 1362
  - diamonds: 37
  - clubs: 48

競り参加率・bid target分布は以前より自然になった一方、♦/♣がほぼ選ばれないスート偏りは極端であり、単なる学習上の好みとして扱わず原因を調査する。

## 目的

「学習済み方策が本当に♦/♣を避ける合理的理由を持っている」のか、「encoding / action ordering / legal mask / rollout / reward / playing policy / diagnostics のどこかにバグや非対称性がある」のかを切り分ける。

このIssueは調査を主目的とし、原因が確定する前に恣意的なスート補正やreward調整は入れない。

## 調査項目

少なくとも以下を確認する。

1. bidding action encoding / decoding
   - PASS + 13〜19 × 4 suits のaction index対応が完全に対称か
   - suit index順とTS/Python/C++間の対応にズレがないか
   - encode/decode round-trip

2. legal action / legal mask
   - 同一bid targetで4 suitsが本当に同条件でlegalになる局面を抽出
   - ♦/♣だけmaskされる、indexが欠落する、別actionへmappingされる等がないか

3. observation encoding
   - 4 suitsのhand/trump関連特徴が対称な定義になっているか
   - suitごとのfeature offset/orderの誤りがないか
   - training/runtimeでorderが一致しているか

4. PPO policy output
   - 学習初期、途中、最終checkpointで suit別logit/probability を比較
   - 同一またはスート置換した観測に対する出力を比較
   - ♦/♣出力が学習前から不自然に抑制されていないか

5. rollout / dataset
   - 実際のbidding sample上でlegalActionsとchosenActionを確認
   - suitDistribution集計そのものに誤りがないか
   - candidateだけでなくFrozen RuleBased / Conservativeのsuit分布も比較

6. environment / game rule
   - suit順位（clubs < diamonds < hearts < spades）による同点raise条件が、合理的にどの程度spades/heartsを優遇するか
   - それだけで今回の 1834/1362/37/48 を説明できるか定量評価

7. Frozen playing policyとの相性
   - `ppo-separated-v1000` のplaying strengthがtrump suit別で極端に異ならないか
   - 同程度の契約条件でspades/heartsとdiamonds/clubsの勝率・terminal rewardを比較
   - playing policy自体が特定suitで弱いことをbidding policyが学んでいる可能性を確認

8. reward / outcome
   - candidate bid suit別に games / Napoleon formation / declaration success / raw reward / learning reward を集計
   - 13/14等targetを揃えた条件でもsuit差が残るか

## 必須の診断出力

可能なら既存run
`~/napoleon_runs/non-playing-ppo-rulebased-conservative-relative-100x500`
を使って、少なくとも以下を比較できる診断コマンド/スクリプトを作る。

- candidate chosen suit count
- legal opportunity count per suit
- chosen / legal opportunity ratio per suit
- target別 × suit別 count
- suit別 declaration success rate
- suit別 mean raw terminal reward
- suit別 mean learning terminal reward
- suit別 Napoleon formation rate
- Frozen opponent側のsuit distribution

さらに、スートを機械的に置換したsynthetic observationでpolicy出力がどう変わるか確認できると望ましい。

## 判断基準

調査結果を次のいずれかへ分類する。

A. 実装バグ
- encoding/mask/order/集計等に非対称性がある
- 修正Issue案を提示する

B. Frozen playing policy由来
- trump suit別playing性能差が主因
- playing policy側の再学習/補正が必要か提案する

C. ルール上の合理的偏り
- suit順位や相手戦略によりspades/hearts優位が定量的に説明できる
- それでも現状の偏りが妥当な範囲か評価する

D. PPO学習上のcollapse/探索不足
- legal opportunityや期待rewardに見合わず特定suitへ収束
- entropy/exploration/initialization等の改善案を提示する

## 非目標

- reward係数の変更
- all-pass payoff変更
- opponent mix変更
- bidding rule変更
- suitごとの人工的bonus/penalty追加
- 原因未特定のままpolicyを再学習
- 親PR #203 のmainへのMerge

## 成果物

- 原因候補ごとの検証結果
- どの層で偏りが発生しているか
- 再現可能な診断コマンド
- 結論（A/B/C/D、複合可）
- 必要なら次の修正Issue案

この調査Issueでは原則コード修正PRは作らず、診断用の小さな追加が必要な場合のみ最小限に留める。

**Comment by eletim** [comment]
## 調査結果: non-playing競りAIのsuit偏り

対象run: `/home/eletim/napoleon_runs/non-playing-ppo-rulebased-conservative-relative-100x500`  
対象worktree: `/home/eletim/napoleon_ws/napoleon-web-issue-202` (`feature/issue-202-non-playing-ai`)

今回は原因調査のみ。reward変更、suit補正、opponent変更、本文コード変更はしていません。worktreeは調査後もcleanです。

### 再現コマンド

Issue本文取得:

```bash
gh issue view 246 --repo eletim/napoleon-web --json number,title,body,state,author,createdAt,updatedAt,labels,comments
```

既存run確認:

```bash
jq '.phases.bidding.rollout.diagnostics.bidding' \
  ~/napoleon_runs/non-playing-ppo-rulebased-conservative-relative-100x500/iterations/iter-000099/iteration.json
```

dataset集計、checkpoint別ONNX推論、synthetic suit置換、Frozen相手再rolloutは、`node - <<'NODE' ... NODE` の一回限り診断で実行。使用した入力はすべて以下のみです。

```bash
cd /home/eletim/napoleon_ws/napoleon-web-issue-202
RUN=~/napoleon_runs/non-playing-ppo-rulebased-conservative-relative-100x500
DATASET=$RUN/iterations/iter-000099/bidding/dataset
# JSONL shardを読み selectedActionIndex/legalBidMask/outcome/rawTerminalReward/terminalReward を集計
# onnxruntime-nodeで $RUN/bootstrap,bidding と iter-000/005/010/020/030/050/070/099 の policy.onnx を batch推論
# runNonPlayingBiddingRlGame({seed:99000202..99000701,candidateSeatIndex:0..4}) で Frozen相手分布と契約suit別結果を再計測
```

検証テスト:

```bash
pnpm --filter @napoleon/ai-observation test -- encodeBiddingAction
pnpm --filter @napoleon/game-core test -- game.test.ts
pnpm --filter @napoleon/cpp-core test -- differential
```

結果: `encodeBiddingAction` 5 tests passed、`game.test.ts` 105 tests passed、C++ core self-test + TS/C++ differential harness passed。

### 1. encoding / order / mask

確認したsuit order:

| 層 | order / 意味 |
| --- | --- |
| TS game bidding priority | `clubs < diamonds < hearts < spades` |
| TS model I/O `BIDDING_HISTORY_SUIT_ORDER` | `spades, hearts, diamonds, clubs` |
| Python constants | `("spades", "hearts", "diamonds", "clubs")` |
| C++ `Suit` enum / playing obs | `Spades=0, Hearts=1, Diamonds=2, Clubs=3` |

`encodeBiddingAction` は `PASS=0`, `1 + (target-13)*4 + suitIndex`。decode round-trip test passed。C++ differentialもpassed。

最終dataset `iter-000099/bidding/dataset`:

```json
{"samples":4370,"pass":1089,"bid":3281,"selectedIllegal":0,"maskLengthBad":0}
```

legal opportunityは♦/♣にも十分あります。mask欠落ではありません。

| suit | chosen | legal opportunity | chosen/legal |
| --- | ---: | ---: | ---: |
| spades | 1834 | 26783 | 0.068476 |
| hearts | 1362 | 25900 | 0.052587 |
| diamonds | 37 | 24870 | 0.001488 |
| clubs | 48 | 23943 | 0.002005 |

結論: A(encoding/mask/order実装バグ)を示す証拠は見つからず。少なくとも「♦/♣だけlegal maskから欠落」「selectedActionIndexが別suitにmapping」は否定的です。

### 2. target別×suit別 candidate選択

| target | spades | hearts | diamonds | clubs |
| ---: | ---: | ---: | ---: | ---: |
| 13 | 973 | 798 | 2 | 3 |
| 14 | 656 | 534 | 14 | 0 |
| 15 | 197 | 20 | 3 | 1 |
| 16 | 0 | 2 | 9 | 23 |
| 17 | 0 | 3 | 2 | 3 |
| 18 | 5 | 4 | 2 | 14 |
| 19 | 3 | 1 | 5 | 4 |

低targetの13/14でほぼ♠/♥だけを選び、♦/♣は高target側に少数だけ残っています。これは単なるsuit順位ルールだけでは説明しづらいです。

### 3. suit別 outcome / reward

candidateがそのsuitを選んだsample単位です。

| suit | bids | Napoleon formation | formation rate | success | success rate | mean raw | mean learning |
| --- | ---: | ---: | ---: | ---: | ---: | ---: | ---: |
| spades | 1834 | 1241 | 0.676663 | 417 | 0.336019 | 6.234460 | 0.578081 |
| hearts | 1362 | 781 | 0.573421 | 293 | 0.375160 | 7.089574 | 0.653157 |
| diamonds | 37 | 30 | 0.810811 | 5 | 0.166667 | 1.729730 | -1.313514 |
| clubs | 48 | 47 | 0.979167 | 7 | 0.148936 | 0.541667 | -1.429167 |

♦/♣は選択数が少ない上に、選ばれた時のlearning rewardが負です。ただし分母が小さく、かつ高target寄りなので、これだけで「♦/♣一般が弱い」とは言えません。

### 4. policy logits/probabilitiesの崩れ時期

同じ最終dataset 4370 observationsに各checkpointをbatch推論し、legal mask後softmaxで平均probを集計。

| checkpoint | pass | spades | hearts | diamonds | clubs | greedy pass/S/H/D/C |
| --- | ---: | ---: | ---: | ---: | ---: | --- |
| bootstrap | 0.0412 | 0.2695 | 0.2394 | 0.2268 | 0.2231 | 0/4370/0/0/0 |
| iter000 | 0.0462 | 0.2750 | 0.2363 | 0.2242 | 0.2184 | 5/4365/0/0/0 |
| iter005 | 0.1340 | 0.3641 | 0.2011 | 0.1604 | 0.1404 | 2987/1383/0/0/0 |
| iter010 | 0.3119 | 0.3377 | 0.1662 | 0.1304 | 0.0537 | 4370/0/0/0/0 |
| iter020 | 0.3452 | 0.4632 | 0.1051 | 0.0748 | 0.0116 | 1111/3259/0/0/0 |
| iter030 | 0.3206 | 0.5583 | 0.0806 | 0.0372 | 0.0032 | 1171/3135/64/0/0 |
| iter050 | 0.3060 | 0.4275 | 0.2374 | 0.0267 | 0.0024 | 1229/2076/1065/0/0 |
| iter070 | 0.2634 | 0.4287 | 0.3001 | 0.0046 | 0.0033 | 1124/1889/1357/0/0 |
| iter099 | 0.2562 | 0.4261 | 0.3020 | 0.0062 | 0.0095 | 1145/1891/1334/0/0 |

初期から実装上♦/♣が死んでいるわけではありません。iter5〜20で急速に低suit確率が落ち、後半はpolicyのgreedy出力から♦/♣が消えます。D(PPO collapse/探索不足)の根拠が強いです。

### 5. synthetic suit置換

最終policyに対し、bidding observation内のsuit関連特徴を機械的に置換しました。対象: `selfHandMask`, `legalBidMask`, `highestBidSuitOneHot`, `biddingHistorySuitIndicesOneHot`。

| input variant | pass | spades | hearts | diamonds | clubs | greedy pass/S/H/D/C |
| --- | ---: | ---: | ---: | ---: | ---: | --- |
| original | 0.2562 | 0.4261 | 0.3020 | 0.0062 | 0.0095 | 1145/1891/1334/0/0 |
| swap spades-diamonds | 0.3099 | 0.2841 | 0.3557 | 0.0313 | 0.0190 | 1436/1280/1617/21/16 |
| swap hearts-clubs | 0.3011 | 0.4775 | 0.1925 | 0.0116 | 0.0173 | 1397/2131/837/5/0 |
| reverse | 0.3357 | 0.2459 | 0.3070 | 0.0330 | 0.0783 | 1631/1160/1400/39/140 |

置換しても出力ラベルとして♠/♥が強く残ります。これは実装の固定mappingバグというより、学習済みpolicyがactionラベルに強い非対称バイアスを持った状態です。

### 6. Frozen RuleBased / Conservative側

同じseed範囲・同じcandidate behavior artifact (`iter-000098/bidding`)・同じ`ppo-separated-v1000` playing policyで2500ゲームを再rolloutし、全bidding decisionsを集計。

| actor | pass | bids | spades | hearts | diamonds | clubs |
| --- | ---: | ---: | ---: | ---: | ---: | ---: |
| candidate | 1089 | 3281 | 1834 | 1362 | 37 | 48 |
| RuleBased frozen | 6042 | 3856 | 654 | 971 | 1174 | 1057 |
| Conservative frozen | 9854 | 154 | 11 | 60 | 55 | 28 |

Frozen側は♦/♣が欠落していません。RuleBasedはむしろdiamonds/clubsが多いです。相手戦略が直接♠/♥へ誘導しているというより、candidate policy固有の偏りです。

### 7. `ppo-separated-v1000` trump suit別playing性能

同じ2500ゲームで最後のbidから契約suitを復元し、Napoleon team成功率を集計。

| contract suit | contracts | success rate | avg point cards | avg target | candidate Napoleon | candidate success rate |
| --- | ---: | ---: | ---: | ---: | ---: | ---: |
| spades | 1043 | 0.3816 | 12.362 | 13.919 | 818 | 0.3484 |
| hearts | 877 | 0.4002 | 12.400 | 13.998 | 591 | 0.3858 |
| diamonds | 287 | 0.4774 | 13.986 | 15.084 | 29 | 0.1724 |
| clubs | 293 | 0.3140 | 12.799 | 15.171 | 45 | 0.1333 |

全contractではdiamondsが最も成功率高く、`ppo-separated-v1000` がtrump suitとして一律に♦/♣で弱いとは言いにくいです。一方でcandidate Napoleonだけに絞ると♦/♣の成功率は低いですが、分母が29/45しかなく、candidateが♦/♣を出す時点で高target・特殊な局面に偏っています。

### 8. 結論分類

主分類: **D. PPOの探索不足 / collapse**  
副次要因: **C. ルール上・戦略上の合理的偏りが一部ある**  
現時点では **A. 実装バグ** と **B. Frozen playing policyのsuit別性能差が主因** は支持しません。

根拠:

- encode/decode round-trip、TS game-core、C++ differentialはpass。
- `selectedIllegal=0`, `maskLengthBad=0`。♦/♣ legal opportunityは2.4万件規模で存在。
- bootstrap/iter0ではpolicy確率が4 suitに広く出ており、初期から♦/♣が実装上死んでいない。
- iter5〜20で低suit確率が急落し、後半はgreedy選択から♦/♣が消える。
- Frozen RuleBased/Conservativeは♦/♣を普通に出す。
- `ppo-separated-v1000` は全contract集計ではdiamonds成功率が高く、playing policy全体の低suit性能差だけでは説明できない。
- suit順位ルールは同targetでspades/heartsをraiseしやすくするが、13/14のlegal opportunityに対するchosen率差、synthetic置換後のaction-label偏りまでは説明しきれない。

### 次に修正すべき内容

1. rollout診断を恒久化する
   - target×suit chosen
   - target×suit legal opportunity
   - chosen/legal ratio
   - suit別 formation / success / raw reward / learning reward
   - Frozen opponent policy別suit distribution
   - 最終contract suit別playing outcome

2. PPO側のcollapse対策を検討する
   - entropy coefficient導入またはschedule
   - minimum exploration / temperature schedule
   - suit別action entropy・KLをiterationごとにログ化
   - early iterationで低suitが消え始めた時点のadvantage分布を確認

3. いきなりsuit bonus/penaltyやreward補正は入れない
   - まずadvantageが低suitで本当に負なのか、低sampleによる自己強化なのかを確認する。

4. synthetic suit permutation evalを正式診断にする
   - policyが入力suit置換に対してどれだけaction label固定バイアスを持つかを毎評価で出す。

5. playing policy起因の追加確認
   - controlled contract条件で、同一hand/stateをsuit置換して`ppo-separated-v1000` playingだけ比較する。今回の集計ではB主因は薄いが、完全否定にはcontrolled evalが必要。
